src/plugins/mer/meremulatordevice.cpp

```diff
@@ -454,6 +454,41 @@ MerConnection *MerEmulatorDevice::connection() const
     return m_connection.data();
 }
 
+void MerEmulatorDevice::addPortForwarding(const QString &ruleName, quint16 hostPort,
+                                          quint16 emulatorVmPort) const
+{
+    MerVirtualBoxManager::updatePortForwardingRule(m_connection->virtualMachine(), ruleName,
+                                                   hostPort, emulatorVmPort);
+}
+
+bool MerEmulatorDevice::removePortForwarding(const QString &ruleName)
+{
+    return MerVirtualBoxManager::deletePortForwardingRule(m_connection->virtualMachine(), ruleName);
+}
+
+bool MerEmulatorDevice::hasPortForwarding(quint16 hostPort, QString *ruleName) const
+{
+    QList<QHash<QString, quint16>> rules({MerVirtualBoxManager::fetchVirtualMachineInfo(
+                                                m_connection->virtualMachine()).otherPorts,
+                                            MerVirtualBoxManager::fetchVirtualMachineInfo(
+                                                m_connection->virtualMachine()).qmlLivePorts,
+                                            MerVirtualBoxManager::fetchVirtualMachineInfo(
+                                                m_connection->virtualMachine()).freePorts});
+    for (int i = 0; i < rules.size(); i++) {
+        if (rules[i].values().contains(hostPort)) {
+            if (ruleName)
+                *ruleName = rules[i].key(hostPort);
+            return true;
+        }
+    }
+    return false;
+}
+
+void MerEmulatorDevice::runVirtualMachine() const
```

Comment:
We shouldn't add this. device->connection()->connectTo() should be used instead.